File: Python/LinkedList.py
class Node:

    def __init__(self, data) :
        self.data  = data
        self.next = None

class LinkedList:

    def __init__(self):
        self.head = None
        self.length = 0
    #Functions regarding accessing the value of the linked list

        #Print the whole list
    def print(self):
        head = self.head
        while head!= None:
            print(str(head.data) + " -> ", end = "")
            head = head.next
        print("END")

        #Get the Node at the specified index
    def NodeAtIndex(self, index):
        if index > self.length or index < 0:
            print('Unable to get None. IndexOutOfBounds')
        i = 0
        head = self.head
        while i < index:
            head = head.next
            i+=1
        return head
    
    def get(self, index = 0):
        if index > self.length or index < 0:
            print('Unable to get None. IndexOutOfBounds')
        i = 0
        head = self.head
        while i < index:
            head = head.next
            i+=1
        return head.data
    
    #Functions Regarding Addition Of A Node

        #Add a node to the back of the list
    def append(self, data):
        if self.head == None:
            self.head = Node(data)
        else:
            head = self.head
            while head.next!=None:
                head = head.next
            head.next = Node(data)
        self.length +=1 
    

        #Add a node to the front of the list
    def push(self, data):
        head = self.head
        self.head = Node(data)
        self.head.next = head
        self.length += 1

    
        #Add a node at the specified  index in the list (used 0 based indexing)
    def insertAtIndex(self, index, data):
        if index>self.length or index<0:
            print('Unable to add Node. IndexOutOfBounds')
        i = 0
        head = self.head
        if index==0:
            self.push(data)
            return
        while i<index-1:
            head = head.next
            i+=1
        temp = head.next
        head.next = Node(data)
        head.next.next = temp
        self.length += 1



    #Functions Regarding Removal of A Node

        #Remove the last node from the list and returns it
    def pop(self):
        if(self.length == 0):
            print('Unable to remove Node. IndexOutOfBounds')
            return

        head = self.head
        while head.next!=None:
            head = head.next
        lastNode = head.next
        head.next = None
        self.length -= 1

        return lastNode
    
        #Remove the Node at the specified index (default value = 0)
    def removeAtIndex(self, index = 0):
        if index > self.length or index < 0:
            print('Unable to remove Node. IndexOutOfBounds')
            return 0

        i = 0
        head = self.head
        
        if index ==0 :
            self.head = head.next
            return


        while i < index-1:
            head = head.next
            i+=1

        head.next = head.next.next
        self.length -= 1

    #Functions Regarding Modification of the Linked List
    def reverse(self):
        current = self.head
        prev = None
        while current:
            next = current.next
            current.next = prev
            prev = current
            current = next
        self.head = prev
        
           
            
           
            




def main():
    ll = LinkedList()
    ll.append(1)
    ll.append(2)
    ll.print()
    ll.push(0)
    ll.insertAtIndex(1, -1)
    ll.print()
    print(ll.get(3))


#driver function
if __name__ == '__main__':
    main()

File: Java/LinkedList.java
class LinkedList {
    Node head;
    int length;

    public static void main(String[] args) {
        LinkedList ll = new LinkedList();
        ll.append(1);
        ll.append(2);
        ll.print();
        ll.push(0);
        ll.pop();
        ll.print();
        ll.insertAtIndex(-1, 1);
        ll.removeFromIndex(2);
        ll.print();
        System.out.println(ll.get(0));
    }

    LinkedList() {
        this.head = null;
        this.length = 0;
    }

// Functions regarding accessing the value of the linked list

    // Print the whole list
    void print() {
        Node current = this.head;
        while (current != null) {
            System.out.print(current.data + " -> ");
            current = current.next;
        }
        System.out.println("END");
    }

    //Get the node at the specified index
    Node nodeAtIndex(int index){
        if (index < 0 || index > length) {
            System.out.println("Unable to add Node. IndexOutOfBounds");
            return null;
        }
        Node temp = head;
        for(int i = 0; i<index ; i++){
            temp = temp.next;
        } 
        return temp;
    }

    //Get the data of the node at the specified index
    int get(int index){
        if (index < 0 || index > length) {
            System.out.println("Unable to add Node. IndexOutOfBounds");
            return Integer.MIN_VALUE;
        }
        Node temp = head;
        for(int i = 0; i<index ; i++){
            temp = temp.next;
        } 
        return temp.data;
    }

// Functions Regarding Addition Of A Node

    // Add a node to the back of the list
    void append(int data) {
        if (head == null) {
            head = new Node(data);
            return;
        }
        Node current = this.head;
        while (current.next != null) {
            current = current.next;
        }
        current.next = new Node(data);
        this.length++;
    }

    // Add a node to the front of the list
    void push(int data) {
        Node newNode = new Node(data);
        newNode.next = this.head;
        this.head = newNode;
        this.length++;
    }

// Add a node at the specified index in the list (used 0 based indexing)
    void insertAtIndex(int data, int index) {
        if (index < 0 || index > length) {
            System.out.println("Unable to add Node. IndexOutOfBounds");
            return;
        }
        Node temp = this.head;
        if(index == 0){
            this.push(data);
            return;
        }
        if(index == length-1){
            this.append(data);
            return;
        }
        for (int i = 0; i < index-1; i++) {
            temp = temp.next;
        }
        Node temp1 = temp.next;
        temp.next = new Node(data);
        temp.next.next = temp1;
        this.length++;
    }

// Functions Regarding Removal of A Node

    // Remove the last node from the list and returns it
    Node pop() {
        if (this.length == 0) {
            System.out.println("Unable to remove Node. IndexOutOfBounds");
            return null;
        }

        Node temp = this.head;


        while (temp.next.next != null) {
            temp = temp.next;
        }
        Node lastNode = temp.next;
        temp.next = null;
        return lastNode;
    }

    Node removeFromIndex(int index) {
        if ( index >= this.length || index <0) {
            System.out.println("Unable to remove Node. IndexOutOfBounds");
            return null;
        }
        if (index == 0){
            Node removedNode = head;
            head = head.next;
            return removedNode;
        }
        else if(index == length -1){
            return pop();
        }

        Node temp = this.head;

        for(int i = 0; i<index-1; i++){
            temp = temp.next;
        }
        Node removedNode = temp.next;
        temp.next = temp.next.next;
        return removedNode;
    }

// Functions Regarding Modification of the Linked List

    void reverse() {
        Node current = this.head;
        Node next;
        Node previous = null;

        while (current != null) {
            next = current.next;
            current.next = previous;
            previous = current;
            current = next;
        }
        this.head = previous;
    }
}

class Node {
    Node next;
    int data;

    Node(int value) {
        this.data = value;
        next = null;
    }

}
